sys: new sock submodule for DTLS

pokgak opened the pull request:
### Contribution description

This PR is part of the PR series to introduce a new DTLS sock module in RIOT.

The module introduces a new sock type `sock_dtls`. Similar to the TCP/UDP sock, DTLS sock needs to be implemented separately for each underlying DTLS implementation.

This PR also introduces a new `net_dtls` group in Doxygen to group and document all related DTLS components together for easier discovery.

The example implementation of this PR is provided by #11943.

### Testing procedure

`make static-test` and `make doc` produces no errors or warnings.

### Issues/PRs references

Depends on #11564
Previous discussion #10897

miri64 commented on Makefile.dep:
```diff
@@ -884,6 +884,16 @@ ifneq (,$(filter riotboot_hdr, $(USEMODULE)))
+
+ifneq (,$(filter sock_dtls, $(USEMODULE)))
+    USEMODULE += credman
+    USEMODULE += sock_udp
```
Since DTLS does not only work over UDP, I don't think this should be a dependency.

miri64 replied:
Sorry, forgot about https://github.com/RIOT-OS/RIOT/pull/11909/files#diff-cce49629fe14fa1665fd3fdc748182a4R69 so nvm.

miri64 commented:
`DTLS_PSK` and `DTLS_ECC` should be exposed to the doc somewhere. And maybe renamed `SOCK_DTLS_PSK`/`SOCK_DTLS_ECC`. How about adding

```C
#ifdef DOXYGEN
#define SOCK_DTLS_PSK (0) /**< PSK mode */
#define SOCK_DTLS_ECC (0) /**< ECC mode */
#endif
```

or something similar to `net/sock/dtls.h`? Also, it's not clear to me, if the two are mutually exclusive, if yes it might be better if having a `SOCK_DTLS_MODE` instead that is set to an enumerated list of possible values wouldn't be better.

miri64 commented:
> Using ECC credentials is a bit iffy currently. It works, but takes too long (>10s) to send and receive the data. Probably the same problem as #11859.

Not sure this helps with the problem at hand, but I wanted to point out that there is an implementation guideline for representation of elliptic curves by the light-weigh implementation guidance working group of the IETF: https://tools.ietf.org/html/draft-ietf-lwig-curve-representations-08

miri64 commented on examples/dtls-sock/dtls-client.c:
```diff
@@ -0,0 +1,174 @@
+
+/* TinyDTLS WARNING check */
+#ifdef WITH_RIOT_SOCKETS
+#error TinyDTLS is set to use sockets but the app is configured for socks.
```
Please explain?

pokgak replied:
This is copied from tinydtls example. Tinydtls has flag WITH_RIOT_SOCKET to use socket instead of the sock API but I just noticed this is still a TODO. So removed it for now.

miri64 commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,185 @@
+ *
+ * @return  0 on success
+ * @return  value < 0 on other errors
+ * @return  -EADDRNOTAVAIL, if sock_udp_ep_t::local of @p sock->udp_sock is not
```
`sock_udp_ep_t` does not have a member `local`

pokgak replied:
This supposed to be sock_udp_t. Fixed.

miri64 replied:
Then this is wrong. Because `sock_udp_t` is not guaranteed to have a member `local` either. It has for the GNRC and lwIP implementations, but it is not guaranteed other implementations will have that. Come to think of it `sock->udp_sock` is _also_ implementation specific (in this case to your `tinydtls` implementation. I'd rather either add the local endpoint as an argument or word it more generic, e.g. "[…], if the local endpoint of the UDP sock of @p sock is was not set".

pokgak replied:
> [...] word it more generic, e.g. "[…], if the local endpoint of the UDP sock of @p sock is was not set".

I think this will do. This way there is no need to add to the parameter list, which is quite long already.

miri64 commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,185 @@
+ *
+ * @return The number of bytes received on success
+ * @return value < 0 on error
+ * @return  -EADDRNOTAVAIL, if sock_udp_ep_t::local of @p sock->udp_sock is not
```
Dito

pokgak replied:
Fixed.

pokgak commented:
> `DTLS_PSK` and `DTLS_ECC` should be exposed to the doc somewhere. And maybe renamed `SOCK_DTLS_PSK`/`SOCK_DTLS_ECC`. How about adding
> 
> ```c
> #ifdef DOXYGEN
> #define SOCK_DTLS_PSK (0) /**< PSK mode */
> #define SOCK_DTLS_ECC (0) /**< ECC mode */
> #endif
> ```
> 
> or something similar to `net/sock/dtls.h`?

I think it is better to expose it in tinydtls pkg instead of DTLS sock as `DTLS_PSK` and `DTLS_ECC` is tinydtls specific and other DTLS implementation may not have/use defines to enable support for specific cipher suites (e.g. wolfSSL uses [`wolfSSL_CTX_set_cipher_list()`][1]). There are also other defines that are not documented anywhere other than [examples/dtls-echo README][2] such as DTLS_CONTEXT_MAX and more.

Or would documenting it all in a new doxygen group (e.g. net_encryption) be better? The group could also describe caveats and configuration options of each DTLS implementation and explain how to use credman with DTLS sock. In general, information on how to get started with using (D)TLS in RIOT.

> Also, it's not clear to me, if the two are mutually exclusive

No, it is not mutually exclusive. A DTLS sock should be able to use multiple cipher suites.

[1]: https://www.wolfssl.com/doxygen/group__Setup.html#gac5835a41f08b90828e72ec20d46d6399
[2]: https://github.com/RIOT-OS/RIOT/blob/master/examples/dtls-echo/README.md#handling-the-static-memory-allocation

miri64 commented:
> I think it is better to expose it in tinydtls pkg instead of DTLS sock as `DTLS_PSK` and `DTLS_ECC` is tinydtls specific and other DTLS implementation may not have/use defines to enable support for specific cipher suites (e.g. wolfSSL uses [`wolfSSL_CTX_set_cipher_list()`](https://www.wolfssl.com/doxygen/group__Setup.html#gac5835a41f08b90828e72ec20d46d6399)). There are also other defines that are not documented anywhere other than [examples/dtls-echo README](https://github.com/RIOT-OS/RIOT/blob/master/examples/dtls-echo/README.md#handling-the-static-memory-allocation) such as DTLS_CONTEXT_MAX and more.

> > Also, it's not clear to me, if the two are mutually exclusive
> 
> No, it is not mutually exclusive. A DTLS sock should be able to use multiple cipher suites.

This should not be part of this PR. But how about making these (pseudo) submodules of `tinydtls` then: `tinydtls_psk` and `tinydtls_ecc`. `DTLS_CONTEXT_MAX` seems to be more of a configuration variable.

> Or would documenting it all in a new doxygen group (e.g. net_encryption) be better? The group could also describe caveats and configuration options of each DTLS implementation and explain how to use credman with DTLS sock. In general, information on how to get started with using (D)TLS in RIOT.

I don't see why a new group would be necessary. If those variables are `tinycrypt` specific, they should go into `pkg_tinycrypt` (maybe also exposed to `config`). The doc for the caveats and how to get started would be better suited in a new group `net_dtls` maybe (which may refer to `sock_dtls` and the various implementation, but not group them specifically, see e.g. `net_ipv6` or `net_udp`), the usage of credman with `dtls_sock` should go into `net_dtls_sock` as it is important for its usage.

pokgak commented:
I'm not sure about having tinydtls_psk and tinydtls_ecc as submodules but I'll try it first and open a PR for that first along with the configuration variable documentation and we can continue to discuss it there.

> The doc for the caveats and how to get started would be better suited in a new group `net_dtls` maybe (which may refer to `sock_dtls` and the various implementation, but not group them specifically, see e.g. `net_ipv6` or `net_udp`), the usage of credman with `dtls_sock` should go into `net_dtls_sock` as it is important for its usage.

Yeah this makes more sense. I added the documentation on how to use DTLS sock in net_sock_dtls (33793d7) and general info about DTLS support in RIOT in net_dtls (e5be939).

miri64 commented on pkg/tinydtls/doc.txt:
```diff
@@ -1,7 +1,7 @@
  * @defgroup pkg_tinydtls  TinyDTLS for RIOT
  * @ingroup  pkg
- * @ingroup  net
+ * @ingroup  net net_dtls
```
Interesting to know, that this works!

miri64 commented on examples/dtls-sock/main.c:
```diff
@@ -0,0 +1,53 @@
+int main(void)
+{
+    /* we need a message queue for the thread running the shell in order to
+     * receive potentially fast incoming networking packets */
```
Where is this comment coming from? The message queue is only needed for commands like `ping6` that rely on `gnrc_netapi`. `sock`s as you know rely on `mbox` so don't require the queue.

pokgak replied:
I copied this from the dtls-echo example but other examples such as gnrc_networking, cord_ep, gcoap (cord_ep and gcoap also uses `sock`) also have a similar comment. So maybe it is all copied and pasted from somewhere. Documentation about when the queue is needed somewhere would be nice.

Tested without the queue and it still works. I'll remove the queue.

miri64 replied:
You are right... And technically the comment is also right, just misleading.

The percondition for using `gnrc_netapi` with a thread is documented [here](https://doc.riot-os.org/group__net__gnrc__netreg.html#gafb6a9d0b840aae07c3c1617932835ef1).

miri64 commented on pkg/tinydtls/include/sock_dtls_types.h:
```diff
@@ -0,0 +1,64 @@
+#endif
+
+#ifndef SOCK_DTLS_MBOX_SIZE
+#define SOCK_DTLS_MBOX_SIZE     (8)         /**< Size of mailbox */
```
What mailbox ;-)?

pokgak replied:
Changed to "Size of DTLS sock mailbox".

miri64 commented on pkg/tinydtls/include/sock_dtls_types.h:
```diff
@@ -0,0 +1,64 @@
+ * @brief Information about DTLS sock
+ */
+struct sock_dtls {
+    dtls_context_t *dtls_ctx;               /**< sock context */
```
TinyDTLS context for the sock?

pokgak replied:
Changed.

miri64 commented on pkg/tinydtls/include/sock_dtls_types.h:
```diff
@@ -0,0 +1,64 @@
+    dtls_context_t *dtls_ctx;               /**< sock context */
+    sock_udp_t *udp_sock;                   /**< Underlying UDP sock to use */
+    mbox_t mbox;                            /**< Mailbox for messaging */
+    msg_t mbox_queue[SOCK_DTLS_MBOX_SIZE];  /**< Queue of mailbox */
```
`Queue for struct sock_dtls::mbox` (not sure this is the right notation with non-typedef'd structs, please check the output of `make doc`)

pokgak replied:
Yup, that's correct. Doxygen links correctly to `sock_dtls::mbox`.

miri64 commented on examples/dtls-sock/main.c:
```diff
@@ -38,9 +31,6 @@ static const shell_command_t shell_commands[] = {
 {
-    /* we need a message queue for the thread running the shell in order to
-     * receive potentially fast incoming networking packets */
-    msg_init_queue(_main_msg_queue, MAIN_QUEUE_SIZE);
```
You are still using `gnrc_icmpv6_echo` with `shell_commands` in this example. The corresponding `shell_command` calls `gnrc_netreg_register()` in the main thread, which I [just](https://github.com/RIOT-OS/RIOT/pull/11909#discussion_r309117969) pointed out to you requires the calling thread to have a message queue. 

https://github.com/RIOT-OS/RIOT/blob/dd6a5cde5c84b6deb75cf864e7d70eed6c601461/sys/shell/commands/sc_gnrc_icmpv6_echo.c#L102

 So just removing the queue is the wrong approach.

pokgak replied:
I've split this PR into two and this discussion actually belongs to the new PR, but we can continue here just for this one.

I removed `gnrc_icmpv6_echo` in the latest push. So no `ping6` anymore from the shell for this example. I added it originally for debugging. The example behavior should be the same as the dtls-echo example now. So, in this case, it should be okay, right?

miri64 replied:
> I removed `gnrc_icmpv6_echo` in the latest push. So no `ping6` anymore from the shell for this example. I added it originally for debugging. The example behavior should be the same as the dtls-echo example now. So, in this case, it should be okay, right?

Yepp, under this condition it is correct to remove the queue initialization :-)

pokgak commented:
Addressed the comments and split the PR into two. This PR now introduces the DTLS sock API and a new Doxygen group `net_dtls` to group related DTLS components. The implementation is provided by #11943. Updated the OP.

miri64 commented:
Thanks. I'll have a look later this week.

miri64 commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,557 @@
+ * @brief   Type for a DTLS sock object
+ *
+ * @note    API implementors: `struct sock_dtls` needs to be defined by
+ *          implementation-specific `sock_dtls_types.h`.
```
```suggestion
 *          an implementation-specific `sock_dtls_types.h`.
```

or

```suggestion
 *          the implementation-specific `sock_dtls_types.h`.
```

miri64 commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,557 @@
+
+/**
+ * @brief Information about the established session with
+ *        the remote endpoint. Used when sending and
```
Typically a `@brief` is only one sentence. More information can be added to the `@details` section (which is implicitly created for non-specified paragraphs):

```suggestion
 *        the remote endpoint.
 *
 * This session object is used when sending and receiving data to the endpoint.
```

However, this information is somewhat obvious (as you don't do much else with a `sock`) so maybe the last sentence can be dropped completely.

pokgak replied:
Dropped the last sentence.

miri64 commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,557 @@
+typedef struct sock_dtls_session sock_dtls_session_t;
+
+/**
+ * @brief Must be called once before any other use
```
Exactly once or at least once? Who typically calls it? Why do I need to call it?

pokgak replied:
Added more detailed description.

miri64 commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,557 @@
+ * @brief Creates a new DTLS sock object
+ *
+ * @param[out] sock     The resulting DTLS sock object
+ * @param[in] udp_sock  Existing UDP sock to be used
```
```suggestion
 * @param[in] udp_sock  Existing UDP sock to be used underneath
```

?

miri64 commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,557 @@
+ *
+ * @param[out] sock     The resulting DTLS sock object
+ * @param[in] udp_sock  Existing UDP sock to be used
+ * @param[in] tag       Credential tag of the sock. Used to get the right
```
What does "right" mean in this context?

pokgak replied:
Replaced "right" with a more detailed description ("credential with the same credential tag as the sock").

miri64 commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,557 @@
+ * @return  value < 0 on error
+ */
+int sock_dtls_create(sock_dtls_t *sock, sock_udp_t *udp_sock,
+                     credman_tag_t tag,unsigned method);
```
```suggestion
                     credman_tag_t tag, unsigned method);
```

miri64 commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,557 @@
+ * @param[in] method    Defines the method for the client or server to use.
+ *
+ * @return  0 on success.
+ * @return  value < 0 on error
```
What values? Please specify, it might be important for the user to know.

pokgak replied:
With current tinyDTLS implementation, the function returns only -1 on error. I prefer to maintain `@return < 0 on error` as that can be specified later in the future and not break existing uses of the API.

miri64 replied:
Mh... I'd prefer the API to be as stable as possible, so I really would prefer to define a value here. If `-1` is the go-to value, then this should be documented as such.

pokgak replied:
Set to `-1`.

miri64 commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,557 @@
+ * @param[in] udp_sock  Existing UDP sock to be used
+ * @param[in] tag       Credential tag of the sock. Used to get the right
+ *                      credential from pool.
+ * @param[in] method    Defines the method for the client or server to use.
```
What kind of method are you referring to here?

pokgak replied:
In wolfSSL, there is WOLFSSL_METHOD that refers to the whether TLS or DTLS are going to be used, its version (1.0,  1.1, 1.2) and whether the sock is used as client or server. The TLS/DTLS part is irrelevant for DTLS sock, but the version number and client/server part is still relevant.

I kinda ignored it for tinydtls because it only supports DTLS version 1.2 and there is no difference from the start between a server and client. Every tinydtls endpoint is initially a server and a client is an endpoint that calls `dtls_connect()` later on.

So, I'm not sure what to name the parameter here to describe that "this parameter determines which DTLS version to use and functionality of this sock (client/server)".

miri64 replied:
Maybe rename it version? In any case there need to be some set of defines or enum, that can be given to the API here. Having the value (or at least its variable name) be implementation-dependent would break the abstraction of the API and require the user to always fiddle around in the code if they switch libraries (which is the whole point we try to prevent here).

pokgak replied:
Okay, will do.

miri64 replied:
> In any case there need to be some set of defines or enum, that can be given to the API here.

I just saw, that https://github.com/RIOT-OS/RIOT/issues/10897 mentioned those, so apparently they were in our original plans ;-)

miri64 replied:
I'm referring to

> ```C
> /**
>  * Methods for sock_dtls_create.
>  * Defines the DTLS protocol version to use.
>  */
> #define DTLSv1_SERVER_METHOD
> #define DTLSv1_CLIENT_METHOD
> #define DTLSv12_SERVER_METHOD
> #define DTLSv12_CLIENT_METHOD
> 

```

pokgak replied:
Alternatively, we can separate the DTLS version from role of the sock. For setting the DTLS version, I'm thinking of adding a new function  to configure any other configuration options offered by the underlying DTLS implementation. More on that in this [comment](https://github.com/RIOT-OS/RIOT/pull/11909#issuecomment-519114289)

Aside from the naming, we actually know when creating the DTLS sock, which role (client/server) the sock will take so any function calls required by the DTLS implementation to start a DTLS server can be done directly in `sock_dtls_create()`. Therefore `sock_dtls_init_server()` is not needed anymore and can be removed.

```C
/**
 * @brief DTLS role for the sock
 */
typedef enum {
    SOCK_DTLS_IS_SERVER,      /* DTLS server role */
    SOCK_DTLS_IS_CLIENT,      /* DTLS client role */
} sock_dtls_role_t;

[...]

/**
 * @brief Creates a new DTLS sock object
 *
 * @see net_credman.
 *
 * @param[out] sock     The resulting DTLS sock object
 * @param[in] udp_sock  Existing UDP sock to be used underneath
 * @param[in] tag       Credential tag of the sock. The sock will only use
 *                      credentials with the same tag given here.
 * @param[in] role      DTLS role of the sock.
 *
 * @return  0 on success.
 * @return  -1 on error
 */
int sock_dtls_create(sock_dtls_t *sock, sock_udp_t *udp_sock,
                     credman_tag_t tag, sock_dtls_role_t role);
```

miri64 replied:
> Alternatively, we can separate the DTLS version from role of the sock. For setting the DTLS version, I'm thinking of adding a new function to configure any other configuration options offered by the underlying DTLS implementation. More on that in this [comment](https://github.com/RIOT-OS/RIOT/pull/11909#issuecomment-519114289)

Since the versions are incompatible to each other, I think it is important to cover them in this API, beyond your configuration proposal

pokgak replied:
Added commit to address this. See [comment](https://github.com/RIOT-OS/RIOT/pull/11909#issuecomment-519494942).

pokgak replied:
Added commit to address this. See [comment](https://github.com/RIOT-OS/RIOT/pull/11909#issuecomment-519494942).

miri64 commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,557 @@
+ *
+ * @param[in] sock      DTLS sock to listen to
+ */
+void sock_dtls_init_server(sock_dtls_t *sock);
```
In accordance with `gnrc_tcp`: maybe name it `gnrc_dtls_listen()`? Not sure, if we already talked about this :sweat_smile:...

pokgak replied:
The description might be misleading. There is actually no 'listening' that happens here. What I meant to say is, the function initializes the sock to for any configurations that must be done so that it can be used as a DTLS server.  So maybe:

```C
/**
 * @brief configure the @p sock to be used as a DTLS server.
 *
 * @param[in] sock      DTLS sock
 */
void sock_dtls_init_server(sock_dtls_t *sock);
```

DTLS is actually more like UDP than TCP. There is no specific "listening for new connection/session" function. The listening is done by the receiving function e.g. 'sock_dtls_recv()`. Handshake process are started when a DTLS server received a ClientHello message instead of Application Data from a client.

miri64 replied:
Ok.

miri64 commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,557 @@
+ * @param[out] remote   The established session, cannot be NULL
+ *
+ * @return  0 on success
+ * @return  value < 0 on other errors
```
Are there more than the ones defined below? If yes, define them instead of being vague.

pokgak replied:
No there are not, I'll remove the line.

miri64 commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,557 @@
+void sock_dtls_init_server(sock_dtls_t *sock);
+
+/**
+ * @brief Establish DTLS session to a server. Execute the handshake step in
```
One sentence per `@brief`

pokgak replied:
Fixed.

miri64 commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,557 @@
+ *
+ * @param sock          DTLS sock to destroy
+ */
+void sock_dtls_destroy(sock_dtls_t *sock);
```
```suggestion
void sock_dtls_close(sock_dtls_t *sock);
```

to be in line with the other `sock` APIs?

pokgak replied:
Makes sense. I'll rename it.

miri64 commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,557 @@
+ * @param[in] sock      @ref sock_dtls_t, which the session is established on
+ * @param[in] remote    Remote session to close
+ */
+void sock_dtls_close_session(sock_dtls_t *sock, sock_dtls_session_t *remote);
```
Not to be confused with `sock_dtls_close()` (in case we rename `sock_dtls_destroy()`) how about naming this then `sock_dtls_term_session` or `sock_dtls_terminate_session()` to avoid confusion.

pokgak replied:
`sock_dtls_terminate_session()` sounds good.

miri64 commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,557 @@
+ * @param[in] sock      DTLS sock to use.
+ * @param[out] remote   Remote DTLS session of the received data.
+ *                      Cannot be NULL.
+ * @param[out] data     Buffer where the data should be stored.
```
What kind of data?

pokgak replied:
Changed to use the same description as in UDP sock `sock_udp_recv()`.

miri64 commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,557 @@
+ *                      Cannot be NULL.
+ * @param[out] data     Buffer where the data should be stored.
+ * @param[in] maxlen    Maximum memory available at @p data.
+ * @param[in] timeout   Timeout for receive in microseconds.
```
```suggestion
 * @param[in] timeout   Receive timeout in microseconds.
```

miri64 commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,557 @@
+ *                      May be SOCK_NO_TIMEOUT to wait until data
+ *                      is available.
+ *
+ * @note Function may block if data not available and @p timeout != 0
```
```suggestion
 * @note Function may block if data is not available and @p timeout != 0
```

miri64 commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,557 @@
+ * @param[in] data      Pointer where the data to be send are stored
+ * @param[in] len       Length of @p data to be send
+ *
+ * @note Function may block
```
In which cases does it block?

pokgak replied:
Added detailed description.

miri64 commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,556 @@
+ * Now we can listen to incoming packets using @ref sock_dtls_recv(). The application
+ * waits indefinitely for new packets. If we want to timeout this wait period
+ * we could alternatively set the `timeout` parameter of the function to a value
+ * != @ref SOCK_NO_TIMEOUT. If an error occur we just ignore it and continue looping.
```
```suggestion
 * != @ref SOCK_NO_TIMEOUT. If an error occurs we just ignore it and continue looping.
```

pokgak replied:
Fixed.

miri64 commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,556 @@
+ *
+ * After that, we set the address of the remote endpoint where the packet is
+ * supposed to be sent to and the port the server is listening to which is
+ * DTLS_DEFAULT_PORT(20220).
```
```suggestion
 * DTLS_DEFAULT_PORT (20220).
```

(otherwise it looks like a macro invocation)

pokgak replied:
Fixed.

miri64 commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,556 @@
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ {.c}
+ * sock_udp_ep_t remote;
+ * remote.port = DTLS_DEFAULT_PORT;
+ * remote.netif = gnrc_netif_iter(NULL)->pid;   // only if gnrc_netif_numoff == 1
```
```suggestion
 * remote.netif = gnrc_netif_iter(NULL)->pid;   // only if GNRC_NETIF_NUMOF == 1
```

miri64 replied:
Also (not sure C-style comments are possible here, though): C++-style comment.

pokgak replied:
Fixed.

miri64 commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,556 @@
+ * successfull and the session is established, we can use its `remote` to send
+ * packets to it with @ref sock_dtls_send().
+ *
+ * If error during any of these operation, the session is terminated and the sock
```
"If an error occurs"?

pokgak replied:
Reworded.

miri64 commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,556 @@
+ * res = sock_dtls_establish_session(&dtls_sock, &remote, &session);
+ * if (res < 0) {
+ *     printf("Error establishing session: %d\n", (int)res);
+ *     goto end;
```
Please don't advertise `goto`s in the high-level doc.

pokgak replied:
I'll rewrite this without using `goto`s.

miri64 commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,558 @@
+ * @param[out] sock     The resulting DTLS sock object
+ * @param[in] udp_sock  Existing UDP sock to be used underneath
+ * @param[in] tag       Credential tag of the sock. The sock will only use
+ *                      credentials with the same tag given here.
```
Maybe add a

```
@see net_credman
```

to this functions doc, so people know, what you talk about here ;-).

pokgak replied:
Added link to `net_credman`.

pokgak commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,556 @@
+ *
+ * ## Makefile Includes
+ *
+ * First we need to @ref including-modules "include" a module that implements
```
Changed link style.

pokgak commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,556 @@
+ *
+ * First we need to @ref including-modules "include" a module that implements
+ * this API in our applications Makefile. For example the implementation for
+ * @ref pkg_tinydtls "tinydtls" is called `tinydtls_sock_dtls'.
```
Changed to the second variant.

pokgak commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,556 @@
+ * this API in our applications Makefile. For example the implementation for
+ * @ref pkg_tinydtls "tinydtls" is called `tinydtls_sock_dtls'.
+ *
+ * The corresponding pkg providing the DTLS implementation will be
```
Added link to `pkg`.

pokgak commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,556 @@
+ * manually.
+ *
+ * Each DTLS implementation may have their own configuration options and caveat.
+ * This can be found at @ref net_dtls "DTLS"
```
Removed "DTLS".

pokgak commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,556 @@
+ *
+ * ### Adding credentials
+ *
+ * Before using this module, either as a server or a client, we need to first
```
Changed "module" to "API".

pokgak commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,556 @@
+ *
+ * Before using this module, either as a server or a client, we need to first
+ * add the credentials to be used for the encryption using
+ * @ref net_credman "credman". Note that credman does not copies the credentials
```
Fixed.

pokgak commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,556 @@
+ * Before using this module, either as a server or a client, we need to first
+ * add the credentials to be used for the encryption using
+ * @ref net_credman "credman". Note that credman does not copies the credentials
+ * given into the system, it only have information about the credentials and where
```
Fixed.

pokgak commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,556 @@
+ *
+ * Above we see an example how to register a PSK and a ECC credential.
+ *
+ * First we need to include the header file for @ref net_sock_dtls "DTLS sock".
```
Replaced link to `net_sock_dtls` with "API".

pokgak commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,556 @@
+ *
+ * We tell @ref net_credman "credman" which credential to add by filling in
+ * the credentials information in a struct @ref credman_credential_t. We are
+ * using `CREDMAN_TYPE_PSK` for credman_credential_t::type for PSK credential.
```
Added link to `CREDMAN_TYPE_PSK`. Currently, `make doc` will return
```
warning: unable to resolve reference to `CREDMAN_TYPE_PSK' for \ref command`
```
because the enum values are not documented in `credman.h`. I'll open a PR to fix this in credman.

pokgak replied:
Opened PR #11965

miri64 replied:
#11965 was merged so I believe this needs to be rebased.

pokgak replied:
Rebased. `make doc` returns no warning anymore.

pokgak commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,556 @@
+ * }
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ *
+ * We tell @ref net_credman "credman" which credential to add by filling in
```
Changed.

pokgak commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,556 @@
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ *
+ * After credential infos are filled, we can add it to the credential pool using
+ * `credman_add`.
```
Added link to function.

pokgak commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,556 @@
+ * After credential infos are filled, we can add it to the credential pool using
+ * `credman_add`.
+ *
+ * Adding credentials of other types are all the same to the above except
```
Reworded.

pokgak commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,556 @@
+ * `DTLS_SOCK_SERVER_TAG`, which is defined as `10` beforehand, as our tag.
+ * The last parameter of `sock_dtls_create` is the DTLS version to be used.
+ *
+ * Note that some DTLS implementation does not support earlier versions of DTLS.
```
Fixed.

pokgak commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,556 @@
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ *
+ * Using the initialised UDP sock, we can then create our DTLS sock. We use
+ * `DTLS_SOCK_SERVER_TAG`, which is defined as `10` beforehand, as our tag.
```
Linking to DTLS_SOCK_SERVER_TAG is not possible here because this macro is locally defined in user application and not in `dtls.h`. Reworded the sentence, mentioned that this value is defined in the user application.

pokgak commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,556 @@
+ *
+ * In case of error we stop the program.
+ *
+ * Then we call `sock_dtls_init_server` to initialize the server.
```
Added link.

miri64 commented on sys/include/net/sock/dtls.h:
```diff
@@ -435,6 +437,7 @@ void sock_dtls_init(void);
  * @param[in] udp_sock  Existing UDP sock to be used underneath
  * @param[in] tag       Credential tag of the sock. The sock will only use
  *                      credentials with the same tag given here.
+ *                      @see net_credman.
```
Since this is rendered outside of the parameter doc anyway, I would prefer this to be put before the parameter doc.

miri64 commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,561 @@
+ * }
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ *
+ * After credential informations are filled, we can add it to the credential
```
```suggestion
 * After credential informations are filled, we can add it to the credential
```

miri64 commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,562 @@
+/*
+ * Copyright (C) 2019 HAW Hamburg
```
BTW, just noticed this: I think the authors of the APIs this API is based on  (@danielinux, @rfuentess, @kb2ma) should also keep their copyright to this work. Similar as we included all people planning on the original `sock` into the copyright.

pokgak replied:
Added the copyright and authors. I used names and emails from existing docs I found in RIOT. Not sure that is up to date though.

pokgak commented:
For a more full-featured DTLS implementation like wolfSSL, there are other configs that can be set. Adding a new function for each of the config options will just bloat the API and make it harder to use. Because of that, I'm thinking of adding a new function to the API to configure any options offered by the underlying DTLS implementation. DTLS packages implementing the API can then specify which options are configurable with `sock_dtls_configs_t` and using `sock_dtls_set_configs()` the options are set using the corresponding functions from the DTLS implementation.

```C
/**
 * @brief Optional configuration options for the sock.
 *
 * DTLS implementations might have different degree of configurability.
 * These options should be specified in this struct and set using
 * @ref sock_dtls_set_configs().
 */
typedef struct sock_dtls_configs sock_dtls_configs_t;

/**
 * @brief (Optional) Configure underlying DTLS stack
 *
 * DTLS implementations might have different degree of configurability.
 * These options can be set by calling this function with a filled
 * @ref sock_dtls_configs_t.
 *
 * @param[in] configs    Options to set
 *
 * @return 0 on success
 * @return -1 if at least one of the options if not supported
 */
int sock_dtls_set_configs(sock_dtls_configs_t *configs);
```

miri64 commented:
> For a more full-featured DTLS implementation like wolfSSL, there are other configs that can be set. Adding a new function for each of the config options will just bloat the API and make it harder to use. Because of that, I'm thinking of adding a new function to the API to configure any options offered by the underlying DTLS implementation. DTLS packages implementing the API can then specify which options are configurable with `sock_dtls_configs_t` and using `sock_dtls_set_configs()` the options are set using the corresponding functions from the DTLS implementation.

Mh... But how do you set an instance of `sock_dtls_config_t` stack independently or do you want to leave this out of scope intentionally for now?

pokgak commented:
If I understand your question correctly, for now it could be done using `#ifdef`. A more generic config options that are not stack dependent would be ideal but I don't have any idea for that right now.

```C
sock_dtls_create();

#ifdef MODULE_WOLFSSL_SOCK_DTLS
sock_dtls_config_t config = {
  .cipherlist = "...";
  .version = DTLSv1_2,
};
sock_dtls_set_config(&config);
#endif

sock_dtls_establish_session();
```

pokgak commented:
On second thought, doing it like the example from my last comment doesn't actually solve the problem. It is more straightforward to just use `#ifdef`s and call the DTLS stack function for setting the config option directly.

> For a more full-featured DTLS implementation like wolfSSL, there are other configs that can be set. Adding a new function for each of the config options will just bloat the API and make it harder to use.

But this is still a valid problem, right?

miri64 commented:
> > For a more full-featured DTLS implementation like wolfSSL, there are other configs that can be set. Adding a new function for each of the config options will just bloat the API and make it harder to use.
> 
> But this is still a valid problem, right?

Yeah, I am thinking for a while about something similar to `getsockopt`/`setsockopt` for `sock`, however a clear idea did not crystalize yet (will go in a similar direction as `netopt` though most likely). For now I'd say it is out of the scope of this PR, as their is no general solution for `sock` either.

miri64 commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,603 @@
+    uint8_t minor;  /**< Minor version number */
+} ProtocolVersion;
+
+#define DTLSv1      {254, 255}  /**< DTLS 1.0 version RFC4347#Section-4.1 */
```
```suggestion
/**
 * @brief DTLS v1.0
 * @see [RFC 4347, section 4.1](https://tools.ietf.org/html/rfc4347#section-4.1)
 */
#define DTLSv1      {254, 255}  /**< DTLS 1.0 version RFC4347#Section-4.1 */
```

pokgak replied:
Used numerical values and grouped under `sock_dtls_prot_version` in Doxygen.

miri64 commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,603 @@
+} ProtocolVersion;
+
+#define DTLSv1      {254, 255}  /**< DTLS 1.0 version RFC4347#Section-4.1 */
+#define DTLSv1_2    {254, 253}  /**< DTLS 1.2 version RFC6347#Section-4.1 */
```
Same here.

pokgak replied:
Done.

miri64 commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,603 @@
+typedef struct {
+    uint8_t major;  /**< Major version number */
+    uint8_t minor;  /**< Minor version number */
+} ProtocolVersion;
```
`sock_dtls_prot_version_t`?

miri64 replied:
Regarding your concern in https://github.com/RIOT-OS/RIOT/pull/11909#issuecomment-519494942: names do not matter as long as the positions are right and the name I proposed: a) is more in line with our naming scheme and b) namespaces this type quite clearly.

pokgak replied:
Used numerical values for protocol version, grouped under `sock_dtls_prot_version` in Doxygen.

pokgak commented:
Added some commits. I'll update the higher level documentation if these changes are okay.

(8eb26a9) - The current API does not specify whether the UDP sock used is also closed when `sock_dtls_close()` is called. To avoid calling `sock_udp_close()` twice, I think it should not close the UDP sock. I'm not sure how `sock_udp_close()` behaves if called twice with the same UDP sock but with this, we do not have to worry about this problem.

(8d40d5f) - Added `sock_dtls_method_t` to specify version number and role of the endpoint. Values for DTLS 1.0 and 1.2 are taken from [RFC4347][1] and [RFC6347][2] respectively. `ProtocolVersion` are as define in [RFC5246][3].

[1]: https://tools.ietf.org/html/rfc6347#section-4.1
[2]: https://tools.ietf.org/html/rfc6347#section-4.1
[3]: https://tools.ietf.org/html/rfc5246#section-6.2.1

miri64 commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,603 @@
+    uint8_t minor;  /**< Minor version number */
+} ProtocolVersion;
+
+#define DTLSv1      {254, 255}  /**< DTLS 1.0 version RFC4347#Section-4.1 */
```
How about grouping the different macros?

```C
/**
 * @name  Static version initializers
 * @brief Can be used to initialize @ref sock_dtls_prot_version_t values.
 * @{
 */
...
/** @} */

pokgak replied:
Done.

miri64 commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,603 @@
+#define DTLSv1      {254, 255}  /**< DTLS 1.0 version RFC4347#Section-4.1 */
+#define DTLSv1_2    {254, 253}  /**< DTLS 1.2 version RFC6347#Section-4.1 */
+
+#define DTLS_CLIENT (1),    /**< Endpoint client role */
```
Same here for grouping

```C
/**
 * @name DTLS roles
 * @anchor sock_dtls_roles
 * @{
 */
/** @} */

pokgak replied:
Done.

miri64 commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,603 @@
+ */
+typedef struct {
+    ProtocolVersion version;    /**< DTLS version number */
+    uint8_t role;               /**< Role of the endpoint */
```
You can then link it here:

```suggestion
    /**
     * @brief Role of the endpoint 
     * @see [DTLS roles](@ref sock_dtls_roles))
     */
    uint8_t role;
```

pokgak replied:
Done.

miri64 commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,603 @@
+} ProtocolVersion;
+
+#define DTLSv1      {254, 255}  /**< DTLS 1.0 version RFC4347#Section-4.1 */
+#define DTLSv1_2    {254, 253}  /**< DTLS 1.2 version RFC6347#Section-4.1 */
```
For [DTLS v1.3](https://tools.ietf.org/html/draft-ietf-tls-dtls13-32) there seems to be no version number defined (see section 4), so there needs to be a different way to convey the usage of that version to the API. How about we just use numerical values to define the version? This would also safe RAM when allocating `gnrc_dtls_method_t`.

pokgak replied:
It is actually defined [here][1] under `legacy_version` in the last sentence but anyways, it is not that important for the API as we are only using this internally and not with other nodes. I'll replace this with numerical values and group it together as you've shown in the other comments.

[1]: https://tools.ietf.org/html/draft-ietf-tls-dtls13-32#section-5.3

miri64 replied:
Mh... that value is identical to `DTLSv1_2` so I guess going for numerical values is indeed better ;-)

pokgak replied:
Done.

miri64 commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,618 @@
+ */
+#define DTLSv1_0    (1)    /**< DTLS version 1.0 */
+#define DTLSv1_2    (2)    /**< DTLS version 1.2 */
+/** @} */
```
I think it is safe to already add `SOCK_DTLS_1_3`. AFAIK wolfSSL supports it already ;-).

pokgak replied:
Added.

miri64 commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,618 @@
+ * @anchor sock_dtls_prot_version
+ * @{
+ */
+#define DTLSv1_0    (1)    /**< DTLS version 1.0 */
```
SOCK_DTLS_1_0 to stick to the naming scheme?

pokgak replied:
Renamed.

miri64 commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,618 @@
+ * @anchor sock_dtls_role
+ * @{
+ */
+#define DTLS_CLIENT (1),    /**< Endpoint client role */
```
`SOCK_DTLS_CLIENT`

pokgak replied:
Renamed.

miri64 commented on sys/include/net/sock/dtls.h:
```diff
@@ -452,13 +474,14 @@ void sock_dtls_init(void);
  */
 int sock_dtls_create(sock_dtls_t *sock, sock_udp_t *udp_sock,
-                     credman_tag_t tag, unsigned method);
+                     credman_tag_t tag, sock_dtls_method_t method);
```
This may be premature optimization, but could you (with your `tinydtls` implementation) check how redefining this to

```C
int sock_dtls_create(sock_dtls_t *sock, sock_udp_t *udp_sock,
                     credman_tag_t tag, unsigned version, unsigned role);
```

works out memory-wise? Compared to this? I think from the user experience this would be better, but I'm not sure if this might be more memory-friendly (if it only safes a few bytes or so (<10B), I would still prefer dividing up the values).

pokgak replied:
With `sock_dtls_method_t`, the compiled size of the example application is 20 bytes more compared to the separate version and role. So it is clearly in favor of the separate version. I'll change this.

miri64 commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,618 @@
+ *
+ * @param[in] sock      DTLS sock to listen to
+ */
+void sock_dtls_init_server(sock_dtls_t *sock);
```
Oh, and this is not needed anymore, right?

pokgak replied:
Yup, I'll remove it in the next commit.

pokgak replied:
Done.

miri64 commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,618 @@
+ *          credentials.
+ * @return  -ETIMEDOUT, if timed out when trying to establish session.
+ */
+int sock_dtls_establish_session(sock_dtls_t *sock, sock_udp_ep_t *ep,
```
BTW. https://github.com/RIOT-OS/RIOT/pull/10308 uses the wording `session_create`/`session_destroy` for these functions. It's non-blocking, but I think those names are a bit more succinct.

pokgak replied:
renamed.

miri64 commented:
Great! I think we are getting to a point were we can merge this. I'll have another look soon (hopefully before my vacation next week).

miri64 commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,645 @@
+ *   2. Create DTLS sock @ref sock_dtls_create()
+ *   3. Start listening with @ref sock_dtls_recv()
+ * - Client operation
+ *   1. Create UDP sock @ref sock_udp_create()
```
```suggestion
 *   1. Create UDP sock @ref sock_udp_create() using role `SOCK_DTLS_CLIENT`
```

miri64 commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,645 @@
+ * where it is located at. So it is your responsibility to make sure that the
+ * credential is valid during the lifetime of your application.
+ *
+ * Instruction on how to generate the credentials can be found here [TODO].
```
Maybe remove this sentence for now and re-add it once the doc on how to generate credentials is there.

miri64 commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,645 @@
+ *         res = sock_dtls_recv(&dtls_sock, &session, buf, sizeof(buf),
+ *                              SOCK_NO_TIMEOUT);
+ *         if (res > 0) {
+ *             printf("Received %d bytes: %*.s\n", res, res, buf);
```
This is not a good idea if `buf` is not a string, since then there can be a lot of nasty characters in `buf`.

miri64 replied:
Only if you assume that the client you provide in this example is the one this server is connected with. I don't think its good advertisement to put such a security risk into the server example ;-). People will copy this code and expect it to work.

pokgak replied:
Changed to only print byte length received/sent instead of the content of `buf`.

miri64 commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,645 @@
+ *
+ * To create a DTLS sock, we first need to have an initialized UDP sock.
+ * DTLS sock will inherit the port from given UDP sock. So the port to be used
+ * later to listen to incoming DTLS packets will need to be set here already.
```
This sentence is somewhat hard to parse. Please rephrase.

miri64 commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,645 @@
+ * beforehand, as our tag. Using @ref SOCK_DTLS_1_2 and @ref SOCK_DTLS_SERVER
+ * we set our DTLS endpoint to use DTLS version 1.2 and act as a DTLS server.
+ *
+ * Note that some DTLS implementation does not support earlier versions of DTLS.
```
```suggestion
 * Note that some DTLS implementation do not support earlier versions of DTLS.
```

pokgak replied:
Fixed.

miri64 commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,645 @@
+ *
+ * Now we can listen to incoming packets using @ref sock_dtls_recv(). The
+ * application waits indefinitely for new packets. If we want to timeout this
+ * waiting period we could alternatively set the `timeout` parameter of the
```
```suggestion
 * wait period we could alternatively set the `timeout` parameter of the
```

pokgak replied:
Fixed.

miri64 commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,645 @@
+ *     res = sock_dtls_recv(&dtls_sock, &session, buf, sizeof(buf),
+ *                          SOCK_NO_TIMEOUT);
+ *     if (res > 0) {
+ *         printf("Received %d bytes: %s -- echo!\n", res, buf);
```
Even worse. This could potentially print anything (e.g. secret keys) after `buf` if `buf` is not terminated by a `\0`

pokgak replied:
I'll limit the print to the length of received bytes. More on this in the above comment.

miri64 commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,645 @@
+ *     const char data[] = "HELLO";
+ *     int res = sock_dtls_send(&dtls_sock, &session, data, sizeof(data));
+ *     if (res >= 0) {
+ *         printf("Sent %d bytes: %*.s\n", res, res, data);
```
Here as well.

pokgak replied:
I guess you meant line 355 instead? This line only prints `data` which is a `\0`-terminated string.

miri64 commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,645 @@
+ * }
+ *
+ * if (sock_dtls_session_create(&dtls_sock, &remote, &session) < 0) {
+ *     printf("Error creating session: %d\n", (int)res);
```
`res` is not assigned with the result of `sock_dtls_session_create()`.

pokgak replied:
I missed that. Removed `res`.

miri64 commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,645 @@
+ * @anchor sock_dtls_prot_version
+ * @{
+ */
+#define SOCK_DTLS_1_0    (1)    /**< DTLS version 1.0 */
```
Now that those are numerical, I think an anomymous `enum` is preferable:

```C
enum {
    SOCK_DTLS_1_0 = 1,
    SOCK_DTLS_1_2,
    SOCK_DTLS_1_3,
};
```

pokgak replied:
Changed to `enum`.

miri64 commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,645 @@
+ * @anchor sock_dtls_role
+ * @{
+ */
+#define SOCK_DTLS_CLIENT (1)    /**< Endpoint client role */
```
Here as well.

pokgak replied:
Changed to `enum`.

miri64 commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,645 @@
+typedef struct sock_dtls_session sock_dtls_session_t;
+
+/**
+ * @brief Called exactly once during autoinit.
```
Either "auto-initialization" or "`auto_init`"

pokgak replied:
Changed to `auto_init`.

miri64 commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,645 @@
+ * @param[out] sock     The resulting DTLS sock object
+ * @param[in] udp_sock  Existing UDP sock initialized with
+ *                      @ref sock_udp_create()to be used underneath.
+ * @param[in] tag       Credential tag of the sock. The sock will only use
```
Which sock? To make it clearer that the parameter `sock` is meant, use `@p sock`

pokgak replied:
Fixed.

miri64 commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,645 @@
+ * @param[in] len       Length of @p data to be send
+ *
+ * @note Function may block until a session is created if there is no
+ *       existing session with @p remote.
```
Does this mean, `remote` can change? If so, it should be marked as `@param[in,out]`.

pokgak replied:
No, `remote` does not change. The function only uses the information in `remote ` to create a session with the endpoint pointed to by `remote`.

miri64 replied:
Ok

miri64 commented on sys/include/net/sock/dtls.h:
```diff
@@ -0,0 +1,645 @@
+ * @brief Closes a DTLS sock
+ *
+ * Releases any memory allocated by @ref sock_dtls_create(). This function does
+ * NOT closes the UDP sock used by the DTLS sock. After the call to this
```
```suggestion
 * NOT close the UDP sock used by the DTLS sock. After the call to this
```

pokgak replied:
Fixed.

pokgak commented:
Sorry, Github UI is confusing. I thought I'm replying to the comments. I'll re-post my replies.

miri64 commented on sys/include/net/sock/dtls.h:
```diff
@@ -226,9 +224,9 @@
- * DTLS sock will inherit the port from given UDP sock. So the port to be used
- * later to listen to incoming DTLS packets will need to be set here already.
+ * DTLS sock uses an initialized UDP sock to send and receive encrypted
+ * packets. Therefore, the listening port for the server also need to be
```
needs

pokgak replied:
fixed.

miri64 commented:
Needs another rebase. Other than that, I believe we can finally merge this. You may squash during that rebase.

pokgak commented:
Squashed and rebased.

pokgak commented:
Thanks @miri64  for the review!
